### ISSUE TITLE
Apollo Client 3 reloading query after mutation update

### ISSUE THREAD

**Issue by danfernand** [apollographql/apollo-client#6760]
Hello Apollo team. We have an existing app that we are migrating from Apollo Client 2 to 3. We noticed an issue that is affecting our upgrade. After an apollo mutation modifying an entity that previously a query has loaded we are noticing we are setting loading to true if the fetch policy is set to cache-and-network. 

We use cache-and-network to guarantee that our app is up to date with the backend and expect the cache to be updated with the network silently.

I am not sure if the behavior is expected since its different than AC 2.

Below I have created 2 code sandboxes that are updating a graphql server created with FaunaDB to show the issue.

Apollo Client 3 - https://codesandbox.io/s/funny-https-zb1hf

Apollo Client 2 - https://codesandbox.io/s/ecstatic-cori-t42ce

When updating completed by selecting the opposite value from a todo the whole page is reloading due to using loading from the query to show a loading state.

Both codesanbox examples are using the same graphql server.

I looking all over to confirm that this AC3 behavior is expected but cannot find anything to that regard.

Thank you for your time. 

PS. Great job on the ecosystem! Once we get past this its gonna be awesome!

**Comment by hauxir** [comment]
experiencing the same issue

**Comment by kure-** [comment]
I'm experienced something similar, that might be connected to your behaviour. It feels like an errorneous behaviour of the cache. Your mutation is returning data on the response of the request. The query fetchPolicy is set to `cache-and-network`. The response of the request is automatically taken to write the data to cache and after that write a refetech is triggered and it should not happen. Rather than that, the cache is updated and the component just rerender with fresh data and loading state is not set to true.

The core problem is probably the same in my situation. My mutation is not responding with data. Instead of that, I'm just using simple `refetchQueries` with an array of two query names. The problem here is, that those two queries contain some of the refs of other queries, and those other queries are not part of the refetch array. Those other queries are called within mounted components with queries with fetchPolicy also set to `cache-and-network`.
So what was working in version 2.6.something is, that mutation was called, after it was completed, two network requests have been called, cache has been updated and components were rerendered. The other components, that were using some of the refs of the two queries that were refeteched, were just rerendered with fresh data. No other network requests were made.
The main problem is, that right now, instead of 2 queries that should be refetched, i'm having like 5 or more network requests triggered. Some of them even twice :)

**Comment by benjamn** [comment]
@danfernand One of the things that changed in AC3 is that `cache-and-network` really means to read from the cache _and_ make a network request, not just the first time, but whenever there's a cache update that affects this query.

Although I have a specific recommendation to achieve your desired behavior (verified with your reproduction—thanks!), there's a bit of history here that I think is worth understanding. PR #6221 was a big refactoring (released in `-beta.46`) that made `FetchPolicy` enforcement more consistent, and restored the `-and-network` part of `cache-and-network`, as described above. Realizing that this might not be the behavior folks actually want, we implemented (what we thought was) a clever trick to fall back to `cache-first` after the first network request, in PR #6353. This behavior was still in place when we released 3.0.0, but post-launch feedback convinced us to revert that PR in #6712, because modifying the `options.fetchPolicy` that you requested can be surprising and is not universally appropriate. In other words, your reproduction is making unwanted network requests because `cache-and-network` is being enforced literally.

However, PR #6712 did more than just revert #6353, because `cache-and-network`-followed-by-`cache-first` still seems like a very common hybrid policy that people may want to implement. To make achieving that behavior easy, we also introduced `options.nextFetchPolicy` (available in `@apollo/client@3.1.0`), so you can specify an initial fetch policy (via `options.fetchPolicy`, as before) and also request a different fetch policy next time.

Long story short, if you use `options.nextFetchPolicy` as follows, I believe you can get the behavior you're looking for:
```ts
export default function Todos() {
  const { loading, error, data } = useQuery(GET_ALL_TODOS, {
    fetchPolicy: "cache-and-network",
    nextFetchPolicy: "cache-first",
    variables: {
      _size: 10
    }
  });
  ...
}
```
This means the very first query will read from the cache and make a network request, but subsequent queries will make a network request only if the cache data has become incomplete.

**Comment by danfernand** [comment]
Thank you for your response! Will give nextFetchPolicy a whirl! Apollo has made using graphql amazing. Great job!

**Comment by kure-** [comment]
thanks @benjamn , adding `nextFetchPolicy` seems to be working just well. The behaviour seems to be quite the same as in Apollo version 2 with `fetchPolicy: cache-and-network` :+1:

**Comment by twittwer** [comment]
I added the `nextFetchPolicy: "cache-first"` to a watchQuery.
This watchQuery was updated by a mutation with `optimisticResponse` & `update` function.
After adding the nextFetchPolicy the eatchQuery doesn't emit the optimistic change anymore.

Did someone experience something similar or can point me in the right direction?

**Comment by hinok** [comment]
@twittwer I struggle with the same problem after adding `nextFetchPolicy`.

**Comment by knedev42** [comment]
I have a query that's responsible for fetching a list of entities based on a bunch of variables (filters and pagination).
I'm using the `cache-and-network` fetch policy as I always want to be up-to-date with the backend data and do a refetch when the variables change. It works great, until a mutation happens, that modifies one of the entities in the list...

Since the backend returns the updated entity after the mutation, the cache gets updated automatically, which is awesome.
However an automatic refetch of the query responsible for the list of entities is triggered and there is no other way to prevent that, except by using `nextFetchPolicy: 'cache-first'`. And that's a big problem for me, because I can't use `cache-first`. It makes the component render stale data when switching between different variables (filters/pages) and it stops triggering network requests if there's already data for a certain set of filters & pagination in the cache (the data can be changed by different people at any point and it's important to see the latest version of the entities whenever you change a filter or the page you're on).

I think there should be a way to tell a query using the `cache-and-network` fetch policy to not trigger because of cache updates. Otherwise there are a lot of unnecessary network requests flying and annoying loading states are displayed to the users.
For example something like:
```javascript
useQuery(SOME_QUERY, { fetchPolicy: 'cache-and-network', refetchOnCacheUpdates: false });
```

**Comment by sammarks** [comment]
Just as an FYI, I think the change in the PRs mentioned by @benjamn introduced a _pretty nasty_ bug for me (running v3.1.3): if I used a hook in two different components, it would result in those queries constantly pinging the server for new data (10 requests per second or so, as fast as the browser / network could go it seemed).

The solution mentioned also worked, I just set the default `watchQuery` options as follows:

```js
defaultOptions: {
  watchQuery: {
    fetchPolicy: 'cache-and-network',
    nextFetchPolicy: 'cache-first',
  },
},
```

It still seems to be making more requests than it should (so I probably just need to tweak the policies I selected in the defaults above to fix that), but at least it's not sending the same query to the server 10x every second.

**Comment by edbond88** [comment]
@benjamn  I think it would be nice to add this example to the documentation in the section about the update

**Comment by tim-field** [comment]
This should really be part of the migration guide 🙏 Went down a lot of rabbit wholes to find this issue! https://www.apollographql.com/docs/react/migrating/apollo-client-3-migration/

**Comment by danfernand** [comment]
We have optimistic responses in our apps and they did not change at all. Two things that I can think of is optimistic payload is incorrect and thus cache is not being updated. There should be an error on the console for this. Another thing is that Apollo client 3 cache cannot cache object if there is no _id or id fields. There would be a console info for this!

**Comment by twittwer** [comment]
@danfernand The optimistic responses are not changing. The problem I tried to explain is about the active watchQueries. They are not updated after a mutation with optimisticResponse+updateFunction. The watchQuery with `fetchPolicy: "cache-and-network"` gets updated, but as soon as I add `nextFetchPolicy: "cache-first"` it doesn't get any update from optimistic mutations.

**Comment by esseswann** [comment]
I have the following situation:
```graphql
query A {
  rootField {
     field {
        subfield {
          ...
       }
    }
  }
}
```
where `subfield` is an array
Mutation that returns non of the `filed` or `subfield`
Update function that removes items from `subfield` array using `writeFragment` to `field`
When I call writeFragment the query gets updated. I added `cache-first` and it still gets updated. I tried using `cache-only` but the `{ data }` prop becomes empty, I suppose because Apollo thinks that the cache is incomplete
I would like to only modify the `subfield` array without causing any extra requests

**Comment by blasterpistol** [comment]
@hinok @jperl Same problem, I'm currently using a custom wrapper to bring the behavior of AC 2, but it shouldn't be:
```ts
export function useQuery<Data, Vars>(
  query: DocumentNode,
  options: Omit<QueryHookOptions<Data, Vars>, 'fetchPolicy'>,
) {
  /* make a new request only if the variables have changed
     it's not the same as cache-and-network because cache-and-network refetches 
     the request after changes made by other queries and mutations
     all combinations of fetchPolicy and nextFetchPolicy also not work as in AC2 */

  // track vars changes
  const vars = useRef<Vars | undefined>(undefined)
  const fetchPolicy =
    vars.current === undefined // just mounted so show data from the cache and do request to update data
      ? 'cache-and-network'
      : isEqual(vars.current, options.variables) // check if vars have changed
      ? 'cache-first' // if not so do not make new requests (this is an actual fix)
      : 'network-only' // if so made a request and write it to cache

  const r = useQuery<Data, Vars>(query, { ...options, fetchPolicy })

  // optionally prevent data flickering when vars changing
  const data = useRef<Data | undefined>()
  data.current = r.data || data.current

  return { ...r, data: data.current }
}
```

**Comment by laij84** [comment]
I just upgraded from 3.0.2 to 3.2.1 and am experiencing the same issue, for both `cache-and-network` and `network-only` fetch policies.

**Comment by FedeBev** [comment]
I've been working on a problem similar to this for days.
I have an infinite-scoll list, for that reason I use the `fetchMore` function to incrementally load the required data; in my case this behavior occurs after the `fetchMore` function is called.
The final result is that the number of the query made to the server is duplicated.

Based on what @benjamn said, I guess that this happens because of the cache update after the new data loaded with `fetchMore` are merged with the existing ones.

Setting `fetchPolicy: "cache-and-network"` and  `nextFetchPolicy: "cache-first"` solved the issue.

I definitely agree with @tim-field: this should really be part of the documentation and, in my opinion, the migration guide there should have paragraph to make users aware.

**Comment by Momics** [comment]
I have an issue related to this, but I can't find a nice way of tackling it with fetchPolicy and nextFetchPolicy. My situation is as follows:

I have a list of paginated 'posts' that users can 'like'. This list is ordered by updatedAt. Posts can be updated by any users, so I always need to fetch the latest page of posts to ensure the client is up to date. However, what happens when using the folowing config:

```js
const { data, loading, fetchMore } = usePostsQuery({
    variables: {
      first: pageSize,
      after: undefined,
    },

    fetchPolicy: "cache-and-network",
    nextFetchPolicy: 'cache-first',
    notifyOnNetworkStatusChange: true,
});
```

Is that any time the user fetches a new page that contained no new posts. That request is cached and requests that follow never fetch new pages. So to solve that I changed it to using cache-and-network for both:

```js
const { data, loading, fetchMore } = usePostsQuery({
    variables: {
      first: pageSize,
      after: undefined,
    },

    fetchPolicy: "cache-and-network",
    nextFetchPolicy: 'cache-and-network',
    notifyOnNetworkStatusChange: true,
});
```

However, that introduces a new (mostly UX) issue... Namely, whenever a user mutates a post (e.g. through a 'like') the backend returns the updated post and Apollo Client invalidates the cache, causing the pagination query to refetch. The updated post now appears in the new page and then it gets merged to the top of the list. This is normally desired behaviour when fetching the latest posts. However, when a user is scrolled down on the page and the user likes a post, the post dissapears from view (goes to top). I'd like Apollo to mutate it in place.

**So in short**
I'd like to either have Apollo Client not refetch my paginated query when the user mutates and invalidates its query. Or I'd like an option to bypass the nextFetchPolicy when using fetchMore, so that I am certain I will fetch a new page from the backend and not from cache.

**Comment by FedeBev** [comment]
@Momics, Have you implemented a custom `InMemoryCache` field resolver for you query?
If you haven't, have a look at the [documentation about how to customize the cache behavior](https://www.apollographql.com/docs/react/caching/cache-field-behavior/): with a custom merge function you should be able to keep the order you want of your data.
For example you can avoid to invalidate the whole cache and change only the updated objects as desired.

(OT TIP: In my opinion, subscriptions are the best choices for use cases like that)

**Comment by Momics** [comment]
@FedeBev Thanks for your quick reply!

Indeed I have a field resolver for my query and you made me look at it again which is great. I found that my endCursor and startCursor weren't updating properly, they were both being set to null when getting an empty response from the server. This resulted in the fetchMore function fetching the wrong pages. 🤕

Fetchmore is now working properly even with nextFetchPolicy set to cache-first.

And thanks for the tip, I do have subscriptions aswell. I just left them out of the question because they weren't affecting the issue.

**Comment by ostrebler** [comment]
Same problem here with `network-only`. Setting `nextFetchPolicy` to `cache-first` doesn't solve the problem : it stops from refetching "parent" data, but in fact it stops all refetching, even if variables change.

**Comment by alamothe** [comment]
This is definitely a migration blocker. `nextFetchPolicy` does not solve the issue as it blocks all further updates. I am very worried about all the side effects of `nextFetchPolicy`. 

Currently it is unclear how to achieve the correct (Apollo V2) behavior.

**Comment by nyrf** [comment]
Seems `nextFetchPolicy` can work

**Comment by jakubnavratil** [comment]
Documentation really lacks information about this. There is not even mention about fact, that running one query can trigger another query refetch - . This is very unexpected behavior, can cause infinite loops https://stackblitz.com/edit/apollo-watch-query-infinite-zhiw6w?file=index.js and should be well document with DANGER warnings.

I understood, that fetching entity can update cache and this cache update updates other queries - but without network request. At least this was 2.0 behavior.
Now, with this new behavior (which again, should be clearly stated in docs), how can I manually refetch from network when I have to set `nextFetchPolicy` to `cache-first`?

**Comment by FedeBev** [comment]
@jakubnavratil I totally agree with you.

I guess the only way to refetch from network is to call the `refetch` function returned by the `useQuery` hook. Of course the fetchPolicy should allow network requests, therefore it can't be `cache-only`

**Comment by jakubnavratil** [comment]
I tested it and manual `refetch` works ok with `nextFetchPolicy: cache-first`. I guess lack of documentation and unfortunate name `nextFetchPolicy` causes lot of misunderstanding.

**Comment by sinn1** [comment]
At first `nextFetchPolicy: catch-first` didn't work for me but I found that it was because the fields returned from the mutation didn't match the fields that were specified in the query.

e.g:

```
query {
   todos {
       id
       content
    }
}

mutation addTodo($newTodo: ToDo!) {
   addTodo(todo: $newTodo) {
      content 
   }
}
```

Because response of addToDo mutation doesn't contain `id`, apollo client QueryManager sees the cache as 'incomplete' and reverts to a network query.

You can debug this locally by putting a breakpoint on this line:   https://github.com/apollographql/apollo-client/blob/f9f8bef1ce1e66c4f7cdc2f163db9f345f41791c/src/core/QueryManager.ts#L1102 and reading the error inside `diff`.

**Comment by stangerjm** [comment]
I strongly feel this should be documented. I was trying to get a migration from v2 to v3 working for days when I finally found this thread. I can affirm that using `nextFetchPolicy: cache-first` works.

You can even have this be the default behavior for your queries unless otherwise specified by including it in the `defaultOptions` when building your ApolloClient.

```javascript
export const client = new ApolloClient({
  link,
  cache,
  defaultOptions: {
    watchQuery: {
      fetchPolicy: 'cache-and-network',
      nextFetchPolicy: 'cache-first',
    },
  },
});
```

This should be in the migration guide to get the same behavior as v2. I'm sure it would help a lot of folks correctly migrate to the latest version.

**Comment by FahmyChaabane** [comment]
idk what's going on behind the scenes but, `notifyOnNetworkStatusChange: true,` worked for me.
i had a UseQuery in the most HOC in the app to bring the data of the connected user.. a **refetchQueries** in some mutation inside some child component had the network request running + cache updated.. but the UI wasn't rendering cuz onComplete of the UseQuery ( the one mentioned ) wasn't triggered.

Anyone facing same kind of usecase ?

**Comment by mpgon** [comment]
I am having a problem related to what @sinn1 mentioned above

> At first `nextFetchPolicy: catch-first` didn't work for me but I found that it was because the fields returned from the mutation didn't match the fields that were specified in the query.
> 

This seems like a breaking change not mentioned anywhere? Or maybe I'm missing something. But when upgrading to v3, I have a lot of queries being refetched 2, 3, 4 times because a subsequent query fetches only a subset of the data. But I'm fetching data subsets everywhere. Example:
```
useGetAccount
{
  id
  first_name
  last_name
}
```

```
useGetItem
{
  id
  account {
    id
    first_name
  }
}
```
useGetItem triggers useGetAccount to get fetched again, with "cache-first". only way to solve it is using ""cache-first" + "cache-only" but that's not the behaviour I want when e.g. the variables change
these were never re-triggered in v2
Is there anywhere that documents when a query with "cache-first" is refetched? Couldn't find the rules anywhere.

**Comment by ghun131** [comment]
> I am having a problem related to what @sinn1 mentioned above
> 
> > At first `nextFetchPolicy: catch-first` didn't work for me but I found that it was because the fields returned from the mutation didn't match the fields that were specified in the query.
> 
> This seems like a breaking change not mentioned anywhere? Or maybe I'm missing something. But when upgrading to v3, I have a lot of queries being refetched 2, 3, 4 times because a subsequent query fetches only a subset of the data. But I'm fetching data subsets everywhere. Example:
> 
> ```
> useGetAccount
> {
>   id
>   first_name
>   last_name
> }
> ```
> 
> ```
> useGetItem
> {
>   id
>   account {
>     id
>     first_name
>   }
> }
> ```
> 
> useGetItem triggers useGetAccount to get fetched again, with "cache-first". only way to solve it is using ""cache-first" + "cache-only" but that's not the behaviour I want when e.g. the variables change these were never re-triggered in v2 Is there anywhere that documents when a query with "cache-first" is refetched? Couldn't find the rules anywhere.

I experience the same thing. Queries same `__typename` with different shape tend to trigger one another without any instruction to do so.

**Comment by gourlaa** [comment]
Hi,
We have exactly the same issue for our use case. We were using fetchPolicy: 'cache-and-network' and the useQuery was fired only if the component is mounted on the 2.x version.
Now, when we were changing the sub field with a mutation,we have an unwanted reload of the main query.
The only way we found, it's to add nextFetchPolicy: 'cache-first' locally and add a refetch when we want to bypass the cache-first.
We don't understand why when we are updating the cache with a mutation, other useQuery are fired again ..
It wasn't the case on the 2.x
Best regards,
Pierre

**Comment by dancrew32** [comment]
Huge shoutout to @stangerjm for their proposed fix. Documenting our fix here:

1. Upgraded from ApolloClient V2 to V3
2. Noticed a problem with `react-beautiful-dnd` `onDragEnd` calling a mutation that had `optimisticResponse`, which previously would _not_ show a temporary revert of sort order (correct behavior), suddenly started showing a temporary revert of sort order (incorrect behavior) until the mutation successfully updated the cache.
3. Our `useQuery` hook had an explicit `fetchPolicy: 'network-only'`, because we wanted to avoid showing a stale list.
4. Discovered that setting `fetchPolicy` in the `useQuery` hook, would implicitly set `nextFetchPolicy` with the same policy, but only if there were no `defaultOptions` set in the ApolloClient instance.
5. Tested removing `fetchPolicy: 'network-only'` from `useQuery` and observed correct drag & drop behavior, so this proves that the override was the problem.
6. Set `fetchPolicy: 'network-only' _and_ `nextFetchPolicy: 'cache-first'` on the `useQuery` and observed that the problem was fixed. However, having to set `nextFetchPolicy` on every `useQuery` is impracticable.
7. Thanks to @stangerjm, we tested the following default and now everything is working correctly. We decided to omit `fetchPolicy` as a default from `defaultOptions.watchQuery` so that it uses the framework default of `cache-first`. Only specifying `nextFetchPolicy: 'cache-first'` fixes the problem for us.

```typescript
export const client = new ApolloClient({
  link,
  cache,
  defaultOptions: {
    watchQuery: {
      nextFetchPolicy: 'cache-first',
    },
  },
});
```

**Comment by s123121** [comment]
> I experience the same thing. Queries same `__typename` with different shape tend to trigger one another without any instruction to do so.

As other people have pointed out, using `nextFetchPolicy: cache-first` does not help when the data change shape. So there are more nuances in this problem and this should be fixed inside apollo-client

**Comment by drewlustro** [comment]
Would love to see an update from official maintainers, such as @benjamn, on this matter. The documentation remains incomplete, migration guide makes no mention of fetch policy behavior, and the community cannot help without word about what configuration is necessary to achieve `apollo-client@v2` behavior.

@twittwer's note on optimistic response no longer working under `nextFetchPolicy: 'cache-first'` is particularly of concern.

**Comment by bezreyhan** [comment]
@benjamn just to give an update, upgrading to v3.6.0 did not resolve the issue for me. What's even more problematic for me is that my `onComplete` callback is getting called multiple times.

I currently have to use a `ref` to track if the `onComplete` callback was run. It's not so elegant, but it works. The query is still running multiple times though.

**Comment by efstathiosntonas** [comment]
in my case @sinn1 [comment](https://github.com/apollographql/apollo-client/issues/6760#issuecomment-870904729) helped me a lot debugging the issue of refetch after mutation and cache update. The issue was that the mutation query results keys weren't the same ones as from the query causing a network fetch to properly update the cache. Thanks @sinn1.

**Comment by leymytel** [comment]
@sreubenstone 

I was experiencing the same issue with arrays that had different IDs after an update.

I fixed this issue by adding the same fields in the mutation as the query.

According to the documentation here: https://www.apollographql.com/docs/kotlin/caching/troubleshooting/

> A single missing field in the cache for a query will trigger a network fetch. Sometimes it might be useful to query an extra field early on so that it can be reused by other later queries.

**Comment by sreubenstone** [comment]
@leymytel thanks for the reply. In my case, the `saveUpvote` mutation returns type `Upvote` (I am adding a new Upvote to the upvote array on a specific `Comment` in my news feed). The fields I'm returning off the `saveUpvote` mutation are identical to the field structure for upvotes that come in on Comments when I query `getFeed`. Do I need to return type `Comment` off the `saveUpvote` mutation? I did try this as well (returning the entire Comment, not just the Upvote, on the `saveUpvote` mutation, and it did not work). 

I'm getting lost on the "missing a single field". I have no fields missing that I can see. The upvote returned matches the upvote queried for in getFeed.

```
mutation saveUpvote($comment_id: Int) {
    saveUpvote(comment_id: $comment_id) {
      id
      user_id
      user {
        id
        user_avatar
      }
    }
  }
```

**Comment by alessbell** [comment]
Hi everyone 👋 I've gone ahead and called this out in the migration docs for posterity via https://github.com/apollographql/apollo-client/pull/10714.

I'm going to close this issue out as I see no other actionable items, but please feel free to open a new issue should the need arise. Thanks!
